### PR TITLE
parallel_llvm_nm

### PR DESCRIPTION
**Pull request by juj**
Centralize the use of multiple multiprocessing pools into one to speed up costly pool startup and teardown, and to avoid overallocating pools. Parallelize the reading of linker input files for faster link performance.

(will need testing on Windows before merge, because Windows multiprocessing spawn rules in Python are very different. This was authored on a Linux system)

### CODE REVIEW

File: tools/cache.py [hunk 1]
```diff
@@ -36,7 +36,7 @@ def try_remove_ending(thestring, ending):
 
   def acquire_cache_lock(self):
     if not self.EM_EXCLUSIVE_CACHE_ACCESS:
-      logging.debug('Cache: acquiring multiprocess file lock to Emscripten cache')
+      logging.debug('Cache: PID ' + str(os.getpid()) + ' acquiring multiprocess file lock to Emscripten cache')
```
Comment (kripken) [inline tools/cache.py]: please use `%s` in the string and `% os.getpid()` after it

File: tools/shared.py [hunk 2]
```diff
@@ -895,10 +895,10 @@ def get_vanilla_file():
       logging.debug('failed to use vanilla file, will re-check: ' + str(e))
       is_vanilla = check_vanilla()
     temp_cache = None
+    os.environ['EMCC_WASM_BACKEND'] = str(is_vanilla)
```
Comment (kripken) [inline tools/shared.py]: this change seems unrelated?

File: tools/shared.py [hunk 3]
```diff
@@ -1420,6 +1507,63 @@ def open_make_err(i, mode='r'):
...
+      pool = Building.get_multiprocessing_pool()
+      object_names_in_archives = pool.map(extract_archive_contents, archive_names)
+
+      for n in range(len(archive_names)):
+        Building.ar_contents[archive_names[n]] = object_names_in_archives[n]['files']
+
+      for o in object_names_in_archives:
+        for f in o['files']:
+          if not f in Building.uninternal_nm_cache:
+            object_names += [f]
```
Comment (kripken) [inline tools/shared.py]: `.append(f)`

File: tools/shared.py [hunk 4]
```diff
@@ -1420,6 +1507,63 @@ def open_make_err(i, mode='r'):
...
+    with ToolchainProfiler.profile_block('parallel_llvm_nm'):
+      pool = Building.get_multiprocessing_pool()
+      object_contents = pool.map(g_llvm_nm_uncached, files)
+
+      for i in range(len(files)):
+        Building.uninternal_nm_cache[files[i]] = object_contents[i]
+      return object_contents
+
+  @staticmethod
+  def read_link_inputs(files, just_nm_all_files=False):
```
Comment (kripken) [inline tools/shared.py]: second param seems unused?

File: tools/shared.py [hunk 5]
```diff
@@ -1420,6 +1507,63 @@ def open_make_err(i, mode='r'):
...
+      for n in range(len(archive_names)):
+        Building.ar_contents[archive_names[n]] = object_names_in_archives[n]['files']
+
+      for o in object_names_in_archives:
+        for f in o['files']:
+          if not f in Building.uninternal_nm_cache:
+            object_names += [f]
+
+      # Next, extract symbols from all object files (either standalone or inside archives we just extracted)
+      Building.parallel_llvm_nm(object_names)
```
Comment (kripken) [inline tools/shared.py]: return value seems ignored? and doesn't this method return something? or does it have side effects?

File: tools/shared.py [hunk 6]
```diff
@@ -1168,13 +1168,100 @@ def __setattr__(self, attr, value):
 class Settings(object):
   __metaclass__ = Settings2
 
+# llvm-ar appears to just use basenames inside archives. as a result, files with the same basename
+# will trample each other when we extract them. to help warn of such situations, we warn if there
+# are duplicate entries in the archive
+def warn_if_duplicate_entries(archive_contents, archive_filename_hint=''):
```
Comment (kripken) [inline tools/shared.py]: is this and the next method new code? (new changes to behavior?) or is this just moving code around for multiprocessing purposes? (if so where is the removal?)

**Comment by kripken** [comment]
What kind of speedup does this achieve?

**Comment by juj** [comment]
Thanks for such a quick review! I'll address these once I'm back home next week, working on measuring the effects before landing, and need to test Windows as well.

File: tools/shared.py (hunk 2 above)
Comment (juj) [inline tools/shared.py]: When we spawn child processes in the pool, we want to pass `EMCC_WASM_BACKEND` env var to them (see below), so I wanted to write this out here also in the negative case so that the env var is always present for the child spawns.

File: tools/shared.py (hunk 6 above)
Comment (juj) [inline tools/shared.py]: Oh, good catch, I'm indeed missing the removal, updating in a sec.

**Comment by juj** [comment]
Just did performance benchmarks, and the `calculate_system_libraries` block goes down from 8.8 seconds to 0.8 seconds (-91%) on my PC. I'll get the code reviews addressed in a moment.

File: tools/shared.py (hunk 2 above)
Comment (kripken) [inline tools/shared.py]: Sorry, I don't understand, and I also don't understand where "below" is (below in the file? below in the review comment responses?).

File: tools/cache.py (hunk 1 above)
Comment (juj) [inline tools/cache.py]: ok

File: tools/shared.py (hunk 3 above)
Comment (juj) [inline tools/shared.py]: ok

File: tools/shared.py (hunk 5 above)
Comment (juj) [inline tools/shared.py]: The call is to populate the nm cache for later usage. Adding a comment to denote that the call is for the side effects.

File: tools/shared.py (hunk 4 above)
Comment (juj) [inline tools/shared.py]: Yeah, that was old code, removing that one.

**Comment by juj** [comment]
Alright, updated PR and tested by running through the `other` suite, how does this look like now?

File: tools/shared.py (hunk 2 above)
Comment (juj) [inline tools/shared.py]: Ah sorry, I mean this one: https://github.com/kripken/emscripten/pull/4876/files#diff-b97059c38de7432017d59dee3e87bc72R1261

**Comment by kripken** [comment]
lgtm, thanks.

**Comment by juj** [comment]
Perfect, thanks!
